### PR TITLE
Eng 1572

### PR DESCRIPTION
**Pull request by JonSc8**

### CODE REVIEW

**Review by romCohen — APPROVED** [review]

**Comment by romCohen** [comment]
@JonSc8 please check why the build failed.

File: cloud_utils/cache/utils.py [hunk 1]
```diff
@@ -71,12 +71,9 @@ def _time_since_last_updated(
 )
 
 
-def _get_cache_filename(
-    cache_file_name: str,
-    factory: Callable,
-) -> str:
+def _get_cache_filename(cache_file_name: str, path: str) -> str:
```
Comment (nitzo) [inline cloud_utils/cache/utils.py:74]: this function seems a bit redundant if all it does is join two strings

File: cloud_utils/cache/utils.py [hunk 2]
```diff
@@ -93,42 +90,50 @@ def auto_updating_cache(
     save_local: bool,
     bucket_name: str,
     should_update: Callable[[Optional[datetime.timedelta]], bool],
+    identifier_func: Callable[[Any], str],
+    path: str,
```
[the remark this answers is not in the document]
Comment (nitzo) [reply cloud_utils/cache/utils.py:94]: I think we don't need the function at all and simply send the path from outside.
What we do need is a public function for the default behaviour (save the file in the factory's directory) that can be called from the engine

File: cloud_utils/cache/utils.py [hunk 3]
```diff
@@ -93,42 +90,50 @@ def auto_updating_cache(
     save_local: bool,
     bucket_name: str,
     should_update: Callable[[Optional[datetime.timedelta]], bool],
+    identifier_func: Callable[[Any], str],
```
Comment (nitzo) [inline cloud_utils/cache/utils.py:93]: consider `function_to_identifier`

File: cloud_utils/cache/utils.py [hunk 4]
```diff
@@ -93,42 +90,50 @@ def auto_updating_cache(
...
-        file_store.open_file("r"),
-        json.load,
-        gamla.side_effect(
-            gamla.compose_left(
-                _time_since_last_updated(identifier),
-                _total_hours_since_update,
-                lambda hours_since_last_update: f"Loading cache for [{identifier}]. Last updated {hours_since_last_update} hours ago.",
-                logging.info,
+
+    async def inner(*args):
```
Comment (nitzo) [inline cloud_utils/cache/utils.py:102]: `*args, **kwargs`

File: cloud_utils/cache/utils.py [hunk 5]
```diff
@@ -93,42 +90,50 @@ def auto_updating_cache(
...
-                gamla.side_effect(
-                    _write_to_cache_file(
-                        cache_file,
-                        identifier,
-                        extra_fields,
+            gamla.ternary(
+                gamla.compose_left(_time_since_last_updated(identifier), should_update),
+                gamla.compose_left(
+                    gamla.just(factory),
+                    gamla.apply_async(*args),
```
Comment (nitzo) [inline cloud_utils/cache/utils.py:121]: factory might be sync. so we need to check and use the correct apply, consider adding to gamla (generic apply)

**Review by nitzo — APPROVED** [review]

File: cloud_utils/cache/utils.py (hunk 3 above)
Comment (nitzo) [reply cloud_utils/cache/utils.py:93]: typing also seems wrong

**Pull request merged**